File: fasta_generator.py
import random
from Bio.Seq import Seq
from Bio.SeqRecord import SeqRecord
from Bio import SeqIO

def generate_random_dna_sequence(length):
    bases = ["A", "T", "C", "G"]
    return ''.join(random.choices(bases, k=length))

def save_to_fasta(sequences, file_name):
    fasta_sequences = []
    for i, seq in enumerate(sequences):
        record = SeqRecord(Seq(seq), id=f"sequence_{i+1}", description="")
        fasta_sequences.append(record)
    SeqIO.write(fasta_sequences, file_name, "fasta")

def random_fasta_generator(num_sequences, sequence_length, is_db=False):
    # Generate a list of random DNA sequences
    random.seed(random.randint(1, 1000))
    sequences = [generate_random_dna_sequence(sequence_length) for _ in range(num_sequences)]

    if is_db:
        file_name = f"random_fasta_db_{num_sequences}_{sequence_length}.fasta"
    else:
        file_name = f"random_fasta_{num_sequences}_{sequence_length}.fasta"
    
    save_to_fasta(sequences, file_name)
    print(f"FASTA file saved as {file_name}")

# Generate different fasta db's and fasta files for query
num_sequences_list = [10000, 20000, 50000, 100000]
sequence_variations = 5
for num_sequences in num_sequences_list:
    for sequence_length in range(sequence_variations):
        random_fasta_generator(num_sequences, random.randint(1000, 5000), is_db=True)
        random_fasta_generator(num_sequences, random.randint(1000, 5000), is_db=False)

File: fasta_generator.R
### This file has been replaced by a python script due to low efficiency

# Load necessary library
library(Biostrings)

# Function to generate a random DNA sequence
generate_random_dna_sequence <- function(length) {
  bases <- c("A", "T", "C", "G")
  paste(sample(bases, length, replace = TRUE), collapse = "")
}

# Function to save sequences to a FASTA file
save_to_fasta <- function(sequences, file_name) {
  fasta_content <- ""
  for (i in 1:length(sequences)) {
    fasta_content <- paste(fasta_content, 
                           paste0(">sequence_", i), 
                           sequences[[i]], 
                           sep = "\n")
  }
  write(fasta_content, file = file_name)
}

random_fasta_generator <- function(num_sequences, sequence_length, is_db = F){
  # Generate a list of random DNA sequences
  set.seed(sample(1:1000, 1))
  sequences <- lapply(1:num_sequences, function(x) generate_random_dna_sequence(sequence_length))
  
  if (is_db){
    # Save sequences to a FASTA file
    file_name <- paste0("random_fasta_db_", num_sequences, "_", sequence_length,".fasta")
  }else{
    file_name <- paste0("random_fasta_", num_sequences, "_", sequence_length,".fasta")
  }
    save_to_fasta(sequences, file_name)
    
    cat("FASTA file saved as", file_name)
}
